browser-hub-app/src/client/index.ts

```diff
@@ -0,0 +1,203 @@
+    
+    /** Initialize our websocket connection */
+    const socket = new WebSocket(socketUrl)
+
+    /** Wait for our socket to open successfully */
+    socket.onopen = () => {
+      /** Get public key string */
+      const publicKey = id.public.toString();
+
+      /** Send a new token request */
+      socket.send(JSON.stringify({
+        pub: publicKey,
+        type: 'token'
+      })); 
+
+      /** Listen for messages from the server */
+      socket.onmessage = async (event) => {
+        const data = JSON.parse(event.data)
+        switch (data.type) {
+          /** Error never happen :) */
+          case 'error': {
+            reject(data.value);
+            break;
+          }
+          /** The server issued a new challenge */
+          case 'challenge':{
+            /** Convert the challenge json to a Buffer */
+            const buf = Buffer.from(data.value)
+            /** User our identity to sign the challenge */
+            const signed = await id.sign(buf)
+            /** Send the signed challenge back to the server */
+            socket.send(JSON.stringify({
+              type: 'challenge',
+              sig: signed.toJSON(),
+              pub: publicKey
+            })); 
+            break;
+          }
+          /** New token generated */
+          case 'token': {
+            resolve(data.value)
+            break;
+          }
+        }
+      }
+    }
+  });
+};
+
+class Hub {
+
+  /** The users unique pki identity */
+  id?: Libp2pCryptoIdentity
+
+  /** The Hub API authentication */
+  auth?: UserAuth
+
+  /** Hub API metadata for access control */
+  context: Context = new Context()
+
+  constructor () {}
+
+  setupIdentity = async () => {
+    /** Create or get identity */
+    this.id = await getIdentity();
+    /** Contains the full identity (including private key) */
+    const identity = this.id.toString();
+
+    /** Render our avatar */
+    displayAvatar(identity)
+
+    /** Get the public key */
+    const publicKey = this.id.public.toString();
+
+    /** Display the publicKey short ID */
+    displayIdentity(publicKey)
+  }
+
+  listThreads = async () => {
+    if (!this.auth) {
+      throw Error('User not authenticated')
+    }
+
+    /** Setup a new connection with the API and our user auth */
+    const client = new Client(this.context)
+
+    /** Query for all the user's existing threads (expected none) */
+    const threads = await client.listThreads()
+
+    /** Display the results */
+    displayThreadsList(JSON.stringify(threads));
+  }
+
+  login = async () => {
+    if (!this.id) {
+      throw Error('No user ID found')
+    }
+
+    /** Use the identity to request a new API token */
+    this.auth = await loginWithChallenge(this.id);
+
+    console.log('Verified on Textile API')
+    displayStatus();
+
+    /** Store the access control metadata */
+    this.context = Context.fromUserAuth(this.auth, 'http://localhost:3007')
```

Comment:
Login using the complete auth with user validation (see loginWithChallenge method in same file)